### PR TITLE
feat(backup): L2 constructs

### PR DESCRIPTION
**Pull request by arnulfojr**
closes https://github.com/aws/aws-cdk/issues/4930

- [x] PR title
- [x] unit tests
- [ ] integs
- [ ] Docs

This is my first time using the CDK so feedback is welcome!

## Testing

* Added unit tests according to the standards established by the linting tools.
## Documentation

* Added documentation for the L2 constructs, as established in the contributing guidelines, although some review would be nice to have as well.
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by arnulfojr** [comment]
Seems to be failing due to aws-events not being yet published with version `1.16.1`, I updated my branch to match latest commit in master

**Comment by eladb** [comment]
Hey, this looks like an awesome (and quality) contribution but I am afraid it will take us a while to get back to you with a review since we are a bit busy with re:Invent activities, but this is on our radar and we'll take the time to provide feedback.

Thanks again!

**Comment by arnulfojr** [comment]
hey @eladb let me know when you have an estimated time for a review :)

File: packages/@aws-cdk/aws-backup/README.md [hunk 1]
```diff
@@ -21,3 +21,88 @@ This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aw
...
 ```
+
+### Backup vaults
+
+The place where all backups will be stores is called a vault.
+Create a backup vault for all the backups.
+
+```ts
+new backup.BackupVault(this, 'MyVault', {
+  backupVaultName: 'MyVault',
```
[the remark this answers is not in the document]
Comment (arnulfojr) [reply packages/@aws-cdk/aws-backup/README.md]: The `backupVaultName` is required

File: packages/@aws-cdk/aws-backup/README.md [hunk 2]
```diff
@@ -21,3 +21,88 @@ This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aw
...
+  ],
+});
+```
+
+### Backup selections
+
+```ts
+new backup.BackupSelection(this, 'MySelections', {
+  backupPlan: backupPlan,
+  resources: dynamoTables.map(table => table.tableArn),
```
[the remark this answers is not in the document]
Comment (arnulfojr) [reply packages/@aws-cdk/aws-backup/README.md]: I was also wondering about this, I thought of simply making it a `type Resource` and one for `type Resources` as an array but do you have something better for passing the ARNs from different  resources?

File: packages/@aws-cdk/aws-backup/lib/plan.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+  /**
+   * The time in which a backup job has to be completed before being cancelled.
+   * @default backup window is set to start at 5AM UTC and lasts 8 hours.
+   */
+  readonly completionWindow?: Duration;
+  /**
+   * The time after a backup is scheduled before a job is canceled if it doesn't start successfully.
+   * @default backup window is set to start at 5AM UTC and lasts 8 hours.
+   */
+  readonly startWindowAfter?: Duration;
```
[the remark this answers is not in the document]
Comment (arnulfojr) [reply packages/@aws-cdk/aws-backup/lib/plan.ts]: Yea :D I had a tough time finding a different word but this comes directly from the AWS Backup Service API, which is `startWindowMinutes` which one you think it makes more sense? https://docs.aws.amazon.com/aws-backup/latest/devguide/API_BackupRuleInput.html#Backup-Type-BackupRuleInput-StartWindowMinutes

File: packages/@aws-cdk/aws-backup/lib/plan.ts [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+/**
+ * Properties of the Backup Plan.
+ */
+export interface BackupPlanProps {
+  /**
+   * The name of the backup plan
+   */
+  readonly backupPlanName: string;
```
[the remark this answers is not in the document]
Comment (arnulfojr) [reply packages/@aws-cdk/aws-backup/lib/plan.ts]: yea, it can be optional. I'll use the `uniqueId`

**Comment by jogold** [comment]
@arnulfojr are you still working on this? If not, I will be glad to take this over.

**Comment by arnulfojr** [comment]
@jogold hey! good to hear that someone else is also interested on this PR :)
I was thinking of continuing addressing the comments this week or next week.
From your side, is it something urgent? If so I may be able to prio if it's blocking you

**Comment by jogold** [comment]
> From your side, is it something urgent?

Nothing urgent. If you plan to work on this I will add here my suggestions/comments, I think the API can be improved.

**Comment by arnulfojr** [comment]
@jogold sure, comments are welcome!

File: packages/@aws-cdk/aws-backup/lib/index.ts [hunk 5]
```diff
@@ -1,2 +1,6 @@
 // AWS::Backup CloudFormation Resources:
 export * from './backup.generated';
+
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/index.ts]: The L2 `exports` are usually at the top of the file

File: packages/@aws-cdk/aws-backup/lib/vault.ts [hunk 6]
```diff
@@ -0,0 +1,202 @@
+import iam = require('@aws-cdk/aws-iam');
+import kms = require('@aws-cdk/aws-kms');
+import sns = require('@aws-cdk/aws-sns');
+import { Construct, IResource, Resource } from '@aws-cdk/core';
+import { CfnBackupVault } from './backup.generated';
+
+/**
+ * Represents a Backup vault.
+ */
+export interface IBackupVault extends IResource {
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/vault.ts]: I would keep only `backupVaultName` in the `IBackupVault` interface. It's the only property required to deal with a (imported) vault. This will simplify import: you'll only need a `fromBackupVaultName()`

File: packages/@aws-cdk/aws-backup/lib/vault.ts [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+  }
+
+  public readonly backupVaultName: string;
+  public readonly backupVaultArn: string;
+  public readonly encryptionKey?: kms.IKey;
+  public readonly accessPolicy?: iam.PolicyDocument;
+
+  constructor(scope: Construct, id: string, props: BackupVaultProps) {
+    super(scope, id, {
+      physicalName: props.backupVaultName
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/vault.ts]: You can use `this.node.uniqueId` and make `backupVaultName` optional.

File: packages/@aws-cdk/aws-backup/lib/vault.ts [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+  readonly encryptionKey?: kms.IKey;
+  /**
+   * Notification configuration based on SNS Topics.
+   * @default configures no notifications
+   */
+  readonly notifications?: VaultNotifications;
+  /**
+   * Optional access policy for the vault.
+   * @see https://docs.aws.amazon.com/aws-backup/latest/devguide/creating-a-vault-access-policy.html
+   * @default Has not access policy.
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/vault.ts]:
```suggestion
   * @default - access is not restricted.
```

File: packages/@aws-cdk/aws-backup/lib/vault.ts [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+   * Optional KMS Key to encrypt the backups.
+   * @see https://docs.aws.amazon.com/aws-backup/latest/devguide/creating-a-vault.html
+   * @default Backup Service will create one with the alias `aws/backup`.
+   */
+  readonly encryptionKey?: kms.IKey;
+  /**
+   * Notification configuration based on SNS Topics.
+   * @default configures no notifications
+   */
+  readonly notifications?: VaultNotifications;
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/vault.ts]: I would not nest the props here, how about `notificationTopic` and `notificationEvents` as top level props and a default of all events if `notificationTopic` is defined but `notificationsEvents` is not.

(it's not because the L1 nests its props that the L2 should also do it)

File: packages/@aws-cdk/aws-backup/lib/vault.ts [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+  public readonly backupVaultArn: string;
+  public readonly encryptionKey?: kms.IKey;
+  public readonly accessPolicy?: iam.PolicyDocument;
+
+  constructor(scope: Construct, id: string, props: BackupVaultProps) {
+    super(scope, id, {
+      physicalName: props.backupVaultName
+    });
+
+    const accessPolicy = props.accessPolicy;
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/vault.ts]:
```suggestion
    const accessPolicy = props.accessPolicy && props.accessPolicy.toJSON();
```

File: packages/@aws-cdk/aws-backup/lib/vault.ts [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+
+    this.backupVaultName = this.getResourceNameAttribute(resource.ref);
+    this.backupVaultArn = this.getResourceArnAttribute(resource.attrBackupVaultArn, {
+      service: 'backup',
+      resource: 'backup-vault',
+      resourceName: this.backupVaultName,
+    });
+  }
+
+  private renderNotifications(notifications?: VaultNotifications): CfnBackupVault.NotificationObjectTypeProperty | undefined {
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/vault.ts]: You need to `topic.grantPublish(new iam.ServicePrincipal('backup.amazonaws.com'))` somewhere

File: packages/@aws-cdk/aws-backup/package.json [hunk 12]
```diff
@@ -84,10 +84,20 @@
...
+    "@aws-cdk/aws-kms": "1.18.0",
+    "@aws-cdk/aws-sns": "1.18.0"
   },
   "engines": {
     "node": ">= 10.3.0"
   },
+  "awslint": {
+    "exclude": [
+      "awslint:docs-public-apis:@aws-cdk/aws-backup.BackupSelectionProps",
+      "awslint:docs-public-apis:@aws-cdk/aws-backup.VaultNotifications"
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/package.json]: Both are documented, you can drop this

File: packages/@aws-cdk/aws-backup/lib/plan.ts [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+   * The time after a backup is scheduled before a job is canceled if it doesn't start successfully.
+   * @default backup window is set to start at 5AM UTC and lasts 8 hours.
+   */
+  readonly startWindowAfter?: Duration;
+}
+
+/**
+ * Represents a Backup Plan.
+ */
+export interface IBackupPlan extends IResource {
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/plan.ts]: Same as for `IBackupVault`, you only need `backupPlanId` in this interface.

File: packages/@aws-cdk/aws-backup/lib/plan.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-backup/lib/plan.ts]: Using `id` here as a default is fine (+ better for the console experience and length constraint)

File: packages/@aws-cdk/aws-backup/lib/plan.ts [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+   */
+  readonly backupPlanName: string;
+  /**
+   * A list of plan rules to assign to the plan.
+   */
+  readonly rules: PlanRule[];
+  /**
+   * The backup vault where all backup resources will be stored in.
+   */
+  readonly vault: IBackupVault;
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/plan.ts]: Why is this required? Each rule can use its own vault.

File: packages/@aws-cdk/aws-backup/lib/plan.ts [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+   * to a cold storage class.
+   */
+  readonly moveToColdStorageAfter: Duration;
+}
+
+/**
+ * Represents the backup plan rule configuration.
+ * If a vault is provided, this will be used instead of the one from the backup plan.
+ */
+export interface PlanRule {
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/plan.ts]: I would make a `class` of it. This allows to offer nice static methods like `public static daily(backupVault: IBackupVault)`, `public static weekly(backupVault: IBackupVault)`, `public static monthly1Year(backupVault: IBackupVault)`...

File: packages/@aws-cdk/aws-backup/lib/plan.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-backup/lib/plan.ts]:
```suggestion
  readonly startWindow?: Duration;
```

File: packages/@aws-cdk/aws-backup/lib/plan.ts [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+   * The backup vault where all backup resources will be stored in.
+   */
+  readonly vault: IBackupVault;
+}
+
+/**
+ * A Backup Plan specifies where the backup resources will be stored and
+ * what rules will the plan will follow.
+ */
+export class BackupPlan extends BackupPlanBase {
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/plan.ts]: Here also we can have nice static methods that mimic the console experience: `public static dailyWeeklyMonthly5YearRetention(scope: Construct, id: string, backupVault: IBackupVault)` etc.

File: packages/@aws-cdk/aws-backup/lib/plan.ts [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+
+/**
+ * Represents the backup plan rule configuration.
+ * If a vault is provided, this will be used instead of the one from the backup plan.
+ */
+export interface PlanRule {
+  /**
+   * The name of the rule.
+   */
+  readonly ruleName: string;
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/plan.ts]: We can have a default like `${this.node.id}Rule${this.rules.length}`

File: packages/@aws-cdk/aws-backup/lib/plan.ts [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+    }
+
+    const backupRuleResourceTypes: CfnBackupPlan.BackupRuleResourceTypeProperty[] = rules.map(rule => {
+      return {
+        ruleName: rule.ruleName,
+        targetBackupVault: rule.vault ? rule.vault.backupVaultName : this.vault.backupVaultName,
+        completionWindowMinutes: rule.completionWindow ? rule.completionWindow.toMinutes() : undefined,
+        startWindowMinutes: rule.startWindowAfter ? rule.startWindowAfter.toMinutes() : undefined,
+        lifecycle: this.parseRuleLifecycle(rule.lifecycle),
+        scheduleExpression: rule.schedule.expressionString
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/plan.ts]: Not 100% sure here but I think you have to validate that it's a `cron(...)` and not a `rate(...)`

File: packages/@aws-cdk/aws-backup/lib/selection.ts [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+/**
+ * Backup resource selection.
+ */
+export class BackupSelection extends BackupSelectionBase {
+  /**
+   * Imports a backup selection based from the given attributes.
+   */
+  public static fromBackupSelectionAttributes(scope: Construct, id: string, attrs: BackupSelectionAttributes): IBackupSelection {
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/selection.ts]: Not sure we need to be bale to import a backup selection. There's nothing you can do with it once imported.

File: packages/@aws-cdk/aws-backup/lib/selection.ts [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+
+  private parseRole(props: BackupSelectionProps): iam.IRole {
+    if (props.role) {
+      return props.role;
+    }
+
+    const role = new iam.Role(this, 'Role', {
+      assumedBy: new iam.ServicePrincipal('backup.amazonaws.com'),
+      managedPolicies: [
+        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSBackupServiceRolePolicyForBackup')
```
Comment (jogold) [inline packages/@aws-cdk/aws-backup/lib/selection.ts]: Do we need the `AWSBackupServiceRolePolicyForRestores` policy as well?

File: packages/@aws-cdk/aws-backup/lib/selection.ts [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+   * Optional IAM Role to use for selecting and interacting with the AWS resources to backup.
+   * @default IAM Role based from the AWS Backup service.
+   * @see https://docs.aws.amazon.com/aws-backup/latest/devguide/iam-service-roles.html
+   */
+  readonly role?: iam.IRole;
+  /**
+   * A list of ARNs to backup.
+   * @default to no resources.
+   */
+  readonly resources?: string[];
```
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-backup/lib/selection.ts]: https://docs.aws.amazon.com/aws-backup/latest/devguide/whatisbackup.html#supported-resources

[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-backup/lib/selection.ts]: Would be also nice to support tags (`ListOfTags`)

**Comment by MrArnoldPalmer** [comment]
@arnulfojr Hey! Are you still actively working on this? Or can I close it for now and you can reopen when you're ready?

**Comment by jogold** [comment]
I can take this over. @arnulfojr let me know.

**Comment by arnulfojr** [comment]
@jogold @MrArnoldPalmer Hey! sorry I've had little spare time for me to invest in the PR.
I have locally some changes and I'll try to make the test go green and maybe reignite this PR.
Due to the COVID-19 I've been mostly busy at home setting up the some workplace to be able to WFH :D hopefully I should have some time soon-ish.

If somebody needs this ASAP, I can push my local changes as-is (I think the most beneficial would be the updated docs) and someone can pick it up. Let me know :)

**Comment by MrArnoldPalmer** [comment]
There is no rush to finish this but if you want to push your changes for @jogold to pick up feel free.

Otherwise reopen this whenever you're ready to keep moving. I'm gonna close it just to signal we aren't actively working on it right now now.

@jogold feel free to reopen if you pick this up as well.

File: packages/@aws-cdk/aws-backup/lib/plan.ts (hunk 15 above)
Comment (arnulfojr) [reply packages/@aws-cdk/aws-backup/lib/plan.ts]: Just wondering what hour of the day would be the best default to execute the backups?

File: packages/@aws-cdk/aws-backup/lib/plan.ts (hunk 18 above)
Comment (arnulfojr) [reply packages/@aws-cdk/aws-backup/lib/plan.ts]: yes, if I also recall correctly, they only support cron expressions.

**Comment by jogold** [comment]
@arnulfojr I started to work on this. I will try to push my proposal today or tomorrow; we can the discuss/review this together.

**Comment by arnulfojr** [comment]
@jogold :D I'm working in this PR as we speak. I'll push the changes tonight but feel free to propose it

File: packages/@aws-cdk/aws-backup/lib/selection.ts (hunk 20 above)
Comment (arnulfojr) [reply packages/@aws-cdk/aws-backup/lib/selection.ts]: AFAIK that policy is only if a user wants to execute the restores through backup service. Since this implementation is only executing backups I'm a bit skeptical to add the policy, this would prevent unforeseen restores.
Besides that, I don't see any reason why we shouldn't add it. Let me know if you want me to add it

File: packages/@aws-cdk/aws-backup/lib/selection.ts (hunk 19 above)
Comment (arnulfojr) [reply packages/@aws-cdk/aws-backup/lib/selection.ts]: I'll try to delete it, but if I remember properly I had to add it because the lint was complaining that I should add it.

File: packages/@aws-cdk/aws-backup/lib/selection.ts (hunk 20 above)
Comment (arnulfojr) [reply packages/@aws-cdk/aws-backup/lib/selection.ts]: what if we add a method `allowRestore` or similar and we add the policy. this would expose 1. the API and 2. explicit consent from the user to allow this.

**Comment by arnulfojr** [comment]
Hey @eladb @MrArnoldPalmer, let's follow up in @jogold 's PR.
He added support for tags which are really valuable in this service.
